feature-swap-impl/src/main/java/io/novafoundation/nova/feature_swap_impl/domain/validation/validations/SwapRateChangesValidation.kt

```diff
@@ -0,0 +1,34 @@
+package io.novafoundation.nova.feature_swap_impl.domain.validation.validations
+
+import io.novafoundation.nova.common.utils.toPerbill
+import io.novafoundation.nova.common.validation.ValidationStatus
+import io.novafoundation.nova.common.validation.valid
+import io.novafoundation.nova.common.validation.validationError
+import io.novafoundation.nova.feature_swap_api.domain.model.swapRate
+import io.novafoundation.nova.feature_swap_impl.domain.validation.SwapValidation
+import io.novafoundation.nova.feature_swap_impl.domain.validation.SwapValidationFailure
+import io.novafoundation.nova.feature_swap_impl.domain.validation.SwapValidationFailure.NewRateExceededSlippage
+import io.novafoundation.nova.feature_swap_impl.domain.validation.SwapValidationPayload
+import java.math.BigDecimal
+
+class SwapRateChangesValidation(
+    private val getCurrentRate: suspend (SwapValidationPayload) -> BigDecimal,
+) : SwapValidation {
+
+    override suspend fun validate(value: SwapValidationPayload): ValidationStatus<SwapValidationFailure> {
+        val slippage = value.slippage.toPerbill()
+        val selectedRate = value.swapQuote.swapRate()
+        val newRate = getCurrentRate(value)
+        val rateDifference = (selectedRate - newRate).abs() / selectedRate
+        if (rateDifference > slippage.value.toBigDecimal()) {
```

Comment:
More accurate approach is to compare new quote with previosuly known `SwapLimit` (`amountOutMin` or `amountInMax`)

Reply:
I noticed here an another case. Since I use `abs()` I will cause an error even if this swap will be beneficial for user. It seems it needs to be fixed